### DISCUSSION TITLE
[symfony/console] Overriding `Command::run` (instead of proper `Command::execute`) is a trap, how to prevent misuse?

### DISCUSSION THREAD

**Discussion by janedbal** [symfony/symfony#58435] [Q&A]
I'd like to move the discussion from [twitter](https://x.com/janedbal/status/1841443139095011355) here.

I'm willing to send some fix PR, but I'd like to hear suggestions what is the acceptable prevention here? Best fix imo would be to make `execute` method abstract (BC break), but I'd assume that it is not abstract for some (historic?) reason.

cc @OskarStark

**Comment by janedbal** [comment]
Another option would be to make `run` final (BC break).

**Comment by stof** [comment]
`execute` is not abstract because there is an alternative way to register a command: instantiate a Command object and give it a callable through `setCode`

There are special cases that override `run` for legitimate reasons in symfony (`TraceableCommand`, `LazyCommand` and `SingleCommandApplication`), which would not allow making it final.

**Reply by apfelbox** [reply]
Maybe it would make sense to separate a regular command (with abstract `execute`) and add separate CallbackCommand with only `setCode`?

Or even deprecate `setCode` and use anonymous classes?

**Comment by tgalopin** [comment]
Adding a `@final` in phpdoc on `run` could help without breaking BC as it would trigger IDEs/static analysis error detection and Symfony debug ClassLoader would detect it and list it in deprecations. And then in the next major it would be a real final.
